Sources/App/Commands/Analyzer.swift

```diff
@@ -24,48 +24,164 @@ struct AnalyzerCommand: Command {
         }
         return package
     }
 }
+
+
+/// Wrapper around _reconcileVersions to create a non-throwing version (mainly to ensure
+/// that failed futures don't slip through and break the pipeline).
+func reconcileVersions(application: Application, result: Result<Package, Error>) -> EventLoopFuture<(Package, [Version])> {
+    do {
+        let pkg = try result.get()
+        return try _reconcileVersions(application: application, package: pkg)
+            .map { (pkg, $0) }
+    } catch {
+        return application.eventLoopGroup.next().makeFailedFuture(error)
+    }
+}
+
+
+func _reconcileVersions(application: Application, package: Package) throws -> EventLoopFuture<[Version]> {
+    // fetch tags
+    guard let path = Current.fileManager.cacheDirectoryPath(for: package) else {
+        throw AppError.invalidPackageUrl(package.id, package.url)
+    }
+    guard let pkgId = package.id else {
+        throw AppError.genericError(nil, "PANIC: package id nil for package \(package.url)")
+    }
+    let tags: EventLoopFuture<[String]> = application.threadPool.runIfActive(eventLoop: application.eventLoopGroup.next()) {
+        application.logger.info("listing tags for package \(package.url)")
+        let tags = try Current.shell.run(command: .init(string: "git tag"), at: path)
+        return tags.split(separator: "\n").map(String.init)
+    }
+    // FIXME: also save version for default branch (currently only looking at tags)
+
+    // TODO: sas 2020-05-02: is it necessary to reconcile versions or is delete and recreate ok?
+    // It certainly is simpler.
+    // Delete ...
+    let delete = Version.query(on: application.db)
+        .filter(\.$package.$id == pkgId)
+        .delete()
+    // ... and insert versions
+    let insert: EventLoopFuture<[Version]> = tags
+        .flatMapEachThrowing { try Version(package: package, tagName: $0) }
+        .flatMap { versions in
+            versions.create(on: application.db)
+                .map { versions }
+        }
+
+    return delete.flatMap { insert }
+}
+
+
+func getManifest(package: Package, version: Version) -> Result<Manifest, Error> {
+    Result {
+        // check out version in cache directory
+        guard let cacheDir = Current.fileManager.cacheDirectoryPath(for: package) else {
+            throw AppError.invalidPackageUrl(package.id, package.url)
+        }
+        // FIXME: here we'll want to be able to use tag or default branch
+        guard let revision = version.tagName else {
+            throw AppError.invalidRevision(version.id, version.tagName)
+        }
+        try Current.shell.run(command: .gitCheckout(branch: revision), at: cacheDir)
+        let json = try Current.shell.run(command: .init(string: "swift package dump-package"), at: cacheDir)
+        // TODO: sas-2020-05-03: do we need to run tools-version? There's a toolsVersion key in the JSON
+        // Plus it may not be a good substitute for swift versions?
```

Comment:
In the ruby version you call out to `tools-version` but I've seen there's a `toolsVersion` key in the `dump-package` JSON as well (and seems to be populated).

Was there a reason you called `tools-version` or was that just an oversight?

Reply:
In my experience running `tools-version` is not a great way to detect the Swift version of a package. I think it's worth considering whether we even look at the tools version at all. It's certainly giving incorrect Swift versions for some of the packages in the current system.

Maybe we don't run `tools-version` *or* look at that key in this first release and see how bad things are with missing Swift versions? What do you think?

Reply:
Big upvote from me to ditch it for now :)

I've just started crawling packages in earnest and it seems like `.swift-version` files might be a better heuristic - and one that's not in the swiftpm.co, or is it?

Reply:
Yes, `.swift-version` files are going to be a *much* better indicator of version support, and actually, if you look at the to-do list I had it was at the very top of my list for post-release features 😂

<img width="622" alt="Screenshot 2020-05-05 at 09 13 14@2x" src="https://user-images.githubusercontent.com/5180/81046657-c9d83700-8eb0-11ea-870a-2c793a7ad7a7.png">

Reply:
In case it wasn't clear, I'm saying ditch it. It's a bad indicator 👍